In [1]:
"""
Optuna example that optimizes a Keras model
"""
import optuna
from optuna.integration.mlflow import MLflowCallback

from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.backend import clear_session
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import SGD, Adam
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint, ReduceLROnPlateau
from tensorflow.keras import utils
import tensorflow as tf


print(f"TensorFlow {tf.__version__}")
gpus = tf.config.list_physical_devices("GPU")
if gpus:
    print(f"Number of available GPUs : {len(gpus)}")
    tf.config.set_visible_devices(gpus[0],"GPU")
    tf.config.experimental.set_memory_growth(gpus[0],True)
else:
    print("No GPU available, using CPU !!!")    

# To disable GPU use
tf.config.set_visible_devices([], 'GPU')
print("Not using GPU !!!")    


NCONSTIT=32
NFEAT=3
NTRIALS=50
METRIC='categorical_accuracy'
NEPOCHS = 50
BSIZES = [32,64,128]  # batch size for Optuna search
MINLR = 0.0001
MAXLR = 0.0003
MAXNL = 3 # maximum number of layer for Optuna search
MINNEU = 8 # minimum number of neurons per layer for Optuna search
MAXNEU = 256 # maximum number of neurons per layer for Optuna search

# OPTUNA trial syntax
# filters=trial.suggest_categorical("filters", [32, 64]),
# activation=trial.suggest_categorical("activation", ["relu", "linear"]),
# learning_rate = trial.suggest_float("learning_rate", 1e-5, 1e-1, log=True)
# num_hidden = trial.suggest_int("n_units_l{}".format(i), 4, 128, log=True)            
            

def create_model(trial):
    model = Sequential()

    # Define the MLP.
    nlayers = trial.suggest_int("nlayers", 1, MAXNL)
    model = Sequential()
    model.add(Input(shape=(NCONSTIT*NFEAT)))
    for i in range(nlayers):
        nhidden = trial.suggest_int("nhidden_l{}".format(i), MINNEU, MAXNEU, log=True)
        model.add(Dense(nhidden, activation="relu"))
    model.add(Dense(5, activation="softmax"))

    
    # Define the optimizer
    optim = Adam( learning_rate= trial.suggest_float("learning_rate", MINLR, MAXLR, log=True) )

# compile the model
    model.compile(optimizer=optim, loss='categorical_crossentropy', metrics=[METRIC])
    return model


def objective(trial):
    # Clear clutter from previous Keras session graphs.
    clear_session()
    
    # early stopping callback
    es = EarlyStopping(monitor=METRIC, patience=10)

    # Learning rate scheduler 
    ls = ReduceLROnPlateau(monitor=METRIC, factor=0.2, patience=10)


    # model checkpoint callback
    # this saves our model architecture + parameters into bestmodel.h5
    chkp = ModelCheckpoint('bestmodel.h5', monitor=METRIC, verbose=0, save_best_only=True, mode='auto', save_freq='epoch')

#
    model = create_model(trial)
    model.fit(X_train, Y_train, 
              batch_size= trial.suggest_categorical("bsize", BSIZES),  
              epochs=NEPOCHS, 
              verbose=0,
              validation_data=(X_val, Y_val),
              callbacks=[es,ls,chkp],
              shuffle=True,
              )
    # Load the best model
    model = tf.keras.models.load_model( 'bestmodel.h5',)
    
#                                         custom_objects={ "QDense": QDense,"QActivation": QActivation,
#                                                          "QConv1D": QConv1D, "quantized_bits": quantized_bits,
#                                                          "NodeEdgeProjection": NodeEdgeProjection,} )                                        
    
    
    # Evaluate the model on the validation data
    score = model.evaluate(X_test, Y_test, verbose=2)

    return score[1] # return the model accuracy



TensorFlow 2.8.0
Number of available GPUs : 1
Not using GPU !!!


In [2]:
import util.util
import util.data
#from . import util as dsutil


'''
# Load data 
X_test = np.ascontiguousarray( np.load("{}/x_test_{}const.npy".format(datapath, nconst)) )
Y_test = np.load( "{}/y_test_{}const.npy".format(datapath, nconst), allow_pickle=True )
'''



util.util.device_info()
outdir = "./optuna"


kfolds = 5
val_kfold = 4

print("kfold numbet of folds:", kfolds )
print("val fold num:", val_kfold )

train_kfolds = [kfold for kfold in range(kfolds) if kfold != val_kfold]

fpath        =  f'./data_kfold/jets_{NCONSTIT}constituents_ptetaphi_nonorm'
fnames_train = [f'jet_images_c{NCONSTIT}_minpt2.0_ptetaphi_nonorm_{train_kfold}' for train_kfold in train_kfolds]
fname_val    =  f'jet_images_c{NCONSTIT}_minpt2.0_ptetaphi_nonorm_{val_kfold}'


data = util.data.Data.load_kfolds(fpath, fnames_train, fname_val)
print(data.train_data.shape)

X_train = data.train_data
X_val = data.test_data
X_test = data.test_data

Y_train = data.train_target
Y_val = data.test_target
Y_test = data.test_target

# normalization
interquantile_range_32 = [120, 0.27, 0.27]
interquantile_range_16 = [166, 0.24, 0.24]
interquantile_range_8  = [219, 0.20, 0.20]
    
if  NCONSTIT == 8:
    X_train     = X_train / interquantile_range_8
    X_val       = X_val   / interquantile_range_8
    X_test      = X_test  / interquantile_range_8
elif NCONSTIT == 16:
    X_train     = X_train / interquantile_range_16
    X_val       = X_val   / interquantile_range_16
    X_test      = X_test  / interquantile_range_16
elif NCONSTIT == 32:
    X_train     = X_train / interquantile_range_32
    X_val       = X_val   / interquantile_range_32
    X_test      = X_test  / interquantile_range_32
    
    
# Flatten data for MLP input
NINPUT = NCONSTIT * NFEAT
X_train = X_train.reshape(-1, NINPUT)
X_val  = X_val.reshape(-1, NINPUT)
X_test  = X_test.reshape(-1, NINPUT)
    
    
print(X_train.shape, X_test.shape, Y_train.shape, Y_test.shape)



# Set Optuna
study = optuna.create_study(
        study_name="mlp_study",
        sampler=optuna.samplers.TPESampler(),
        pruner=optuna.pruners.HyperbandPruner(),
        direction="maximize",
        storage="sqlite:///optuna_mlp_{}_constit.db".format(NCONSTIT),
        load_if_exists=True,
    )
        
    
    
# Optimize
#study.optimize(objective(data), n_trials=15, gc_after_trial=True)
study.optimize(objective, n_trials=NTRIALS, gc_after_trial=True)


# Results
#trials = study.get_trials(deepcopy=False, states=[TrialState.PRUNED])
trials = study.get_trials(deepcopy=False)
print("Study statistics: ")
print("  Number of finished trials: ", len(study.trials))
print("  Number of trials: ", len(trials))

# 
print("Best trial:")
trial = study.best_trial

print("  Value: ", trial.value)

print("  Params: ")
for key, value in trial.params.items():
    print("    {}: {}".format(key, value))

    
# reload study and continue optimizing
#study = optuna.load_study(study_name="example-study", storage="sqlite:///example.db")
#study.optimize(objective, n_trials=30)    



GPU: METAL
Activating procedural GPU memory growth...
kfold numbet of folds: 5
val fold num: 4

----------------
Data loading complete:
File name: None
Training data size: 480,840
Test data size: 120,210
Number of constituents: 32
Number of features: 3
----------------

(480840, 32, 3)
(480840, 96) (120210, 96) (480840, 5) (120210, 5)


[I 2023-11-03 07:26:03,012] Using an existing study with name 'mlp_study' instead of creating a new one.
2023-11-03 07:26:03.207554: W tensorflow/core/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz


3757/3757 - 1s - loss: 1.2525 - categorical_accuracy: 0.4745 - 791ms/epoch - 211us/step


[I 2023-11-03 07:27:21,509] Trial 2 finished with value: 0.47454455494880676 and parameters: {'nlayers': 1, 'nhidden_l0': 42, 'learning_rate': 0.00011802120724693812, 'bsize': 128}. Best is trial 2 with value: 0.47454455494880676.


3757/3757 - 1s - loss: 1.0213 - categorical_accuracy: 0.5857 - 820ms/epoch - 218us/step


[I 2023-11-03 07:32:34,477] Trial 3 finished with value: 0.5857332944869995 and parameters: {'nlayers': 2, 'nhidden_l0': 38, 'nhidden_l1': 117, 'learning_rate': 0.00023340342926496938, 'bsize': 32}. Best is trial 3 with value: 0.5857332944869995.


3757/3757 - 1s - loss: 1.1043 - categorical_accuracy: 0.5457 - 813ms/epoch - 217us/step


[I 2023-11-03 07:35:34,594] Trial 4 finished with value: 0.5457366108894348 and parameters: {'nlayers': 2, 'nhidden_l0': 22, 'nhidden_l1': 185, 'learning_rate': 0.00016010979717482802, 'bsize': 64}. Best is trial 3 with value: 0.5857332944869995.


3757/3757 - 1s - loss: 1.0360 - categorical_accuracy: 0.5798 - 857ms/epoch - 228us/step


[I 2023-11-03 07:37:38,185] Trial 5 finished with value: 0.579768717288971 and parameters: {'nlayers': 2, 'nhidden_l0': 39, 'nhidden_l1': 204, 'learning_rate': 0.00022745547846333964, 'bsize': 128}. Best is trial 3 with value: 0.5857332944869995.


3757/3757 - 1s - loss: 1.0217 - categorical_accuracy: 0.5873 - 861ms/epoch - 229us/step


[I 2023-11-03 07:39:38,267] Trial 6 finished with value: 0.5872556567192078 and parameters: {'nlayers': 3, 'nhidden_l0': 84, 'nhidden_l1': 15, 'nhidden_l2': 105, 'learning_rate': 0.00013896853898476147, 'bsize': 128}. Best is trial 6 with value: 0.5872556567192078.


3757/3757 - 1s - loss: 1.1163 - categorical_accuracy: 0.5592 - 815ms/epoch - 217us/step


[I 2023-11-03 07:44:27,148] Trial 7 finished with value: 0.5592213869094849 and parameters: {'nlayers': 1, 'nhidden_l0': 107, 'learning_rate': 0.0002540987906022447, 'bsize': 32}. Best is trial 6 with value: 0.5872556567192078.


3757/3757 - 1s - loss: 1.1837 - categorical_accuracy: 0.5141 - 811ms/epoch - 216us/step


[I 2023-11-03 07:47:06,916] Trial 8 finished with value: 0.5140836834907532 and parameters: {'nlayers': 1, 'nhidden_l0': 90, 'learning_rate': 0.00011917150724067958, 'bsize': 64}. Best is trial 6 with value: 0.5872556567192078.


3757/3757 - 1s - loss: 1.2572 - categorical_accuracy: 0.4795 - 770ms/epoch - 205us/step


[I 2023-11-03 07:51:19,484] Trial 9 finished with value: 0.47945261001586914 and parameters: {'nlayers': 1, 'nhidden_l0': 21, 'learning_rate': 0.0002223949755953682, 'bsize': 32}. Best is trial 6 with value: 0.5872556567192078.


3757/3757 - 1s - loss: 1.0944 - categorical_accuracy: 0.5498 - 816ms/epoch - 217us/step


[I 2023-11-03 07:56:35,516] Trial 10 finished with value: 0.549771249294281 and parameters: {'nlayers': 2, 'nhidden_l0': 23, 'nhidden_l1': 248, 'learning_rate': 0.00010935140027694214, 'bsize': 32}. Best is trial 6 with value: 0.5872556567192078.


3757/3757 - 1s - loss: 0.9393 - categorical_accuracy: 0.6285 - 1s/epoch - 266us/step


[I 2023-11-03 07:59:02,924] Trial 11 finished with value: 0.6284834742546082 and parameters: {'nlayers': 3, 'nhidden_l0': 256, 'nhidden_l1': 11, 'nhidden_l2': 120, 'learning_rate': 0.00018147206355438567, 'bsize': 128}. Best is trial 11 with value: 0.6284834742546082.


3757/3757 - 1s - loss: 0.9382 - categorical_accuracy: 0.6275 - 992ms/epoch - 264us/step


[I 2023-11-03 08:01:30,987] Trial 12 finished with value: 0.6275101900100708 and parameters: {'nlayers': 3, 'nhidden_l0': 236, 'nhidden_l1': 10, 'nhidden_l2': 136, 'learning_rate': 0.00018326549896650806, 'bsize': 128}. Best is trial 11 with value: 0.6284834742546082.


3757/3757 - 1s - loss: 0.9360 - categorical_accuracy: 0.6298 - 995ms/epoch - 265us/step


[I 2023-11-03 08:04:01,853] Trial 13 finished with value: 0.6298477649688721 and parameters: {'nlayers': 3, 'nhidden_l0': 252, 'nhidden_l1': 8, 'nhidden_l2': 157, 'learning_rate': 0.00018846971115651093, 'bsize': 128}. Best is trial 13 with value: 0.6298477649688721.


3757/3757 - 1s - loss: 0.9272 - categorical_accuracy: 0.6343 - 991ms/epoch - 264us/step


[I 2023-11-03 08:06:32,076] Trial 14 finished with value: 0.6343066096305847 and parameters: {'nlayers': 3, 'nhidden_l0': 228, 'nhidden_l1': 8, 'nhidden_l2': 255, 'learning_rate': 0.00019005205250409996, 'bsize': 128}. Best is trial 14 with value: 0.6343066096305847.


3757/3757 - 1s - loss: 0.9276 - categorical_accuracy: 0.6329 - 988ms/epoch - 263us/step


[I 2023-11-03 08:09:07,144] Trial 15 finished with value: 0.6328508257865906 and parameters: {'nlayers': 3, 'nhidden_l0': 172, 'nhidden_l1': 24, 'nhidden_l2': 231, 'learning_rate': 0.0002854481525497479, 'bsize': 128}. Best is trial 14 with value: 0.6343066096305847.


3757/3757 - 1s - loss: 0.9255 - categorical_accuracy: 0.6351 - 974ms/epoch - 259us/step


[I 2023-11-03 08:11:46,491] Trial 16 finished with value: 0.635113537311554 and parameters: {'nlayers': 3, 'nhidden_l0': 154, 'nhidden_l1': 24, 'nhidden_l2': 243, 'learning_rate': 0.0002863441027297299, 'bsize': 128}. Best is trial 16 with value: 0.635113537311554.


3757/3757 - 1s - loss: 0.9477 - categorical_accuracy: 0.6253 - 879ms/epoch - 234us/step


[I 2023-11-03 08:15:25,098] Trial 17 finished with value: 0.625305712223053 and parameters: {'nlayers': 3, 'nhidden_l0': 142, 'nhidden_l1': 27, 'nhidden_l2': 22, 'learning_rate': 0.00029934031941985124, 'bsize': 64}. Best is trial 16 with value: 0.635113537311554.


3757/3757 - 1s - loss: 0.9397 - categorical_accuracy: 0.6281 - 1s/epoch - 270us/step


[I 2023-11-03 08:18:31,838] Trial 18 finished with value: 0.6280925273895264 and parameters: {'nlayers': 3, 'nhidden_l0': 155, 'nhidden_l1': 47, 'nhidden_l2': 256, 'learning_rate': 0.00020595804125667717, 'bsize': 128}. Best is trial 16 with value: 0.635113537311554.


3757/3757 - 1s - loss: 1.0288 - categorical_accuracy: 0.5883 - 825ms/epoch - 220us/step


[I 2023-11-03 08:20:13,389] Trial 19 finished with value: 0.5883454084396362 and parameters: {'nlayers': 2, 'nhidden_l0': 71, 'nhidden_l1': 17, 'learning_rate': 0.0002574571570239548, 'bsize': 128}. Best is trial 16 with value: 0.635113537311554.


3757/3757 - 1s - loss: 0.9563 - categorical_accuracy: 0.6220 - 904ms/epoch - 241us/step


[I 2023-11-03 08:23:34,616] Trial 20 finished with value: 0.6219948530197144 and parameters: {'nlayers': 3, 'nhidden_l0': 128, 'nhidden_l1': 8, 'nhidden_l2': 57, 'learning_rate': 0.00020430435602019023, 'bsize': 64}. Best is trial 16 with value: 0.635113537311554.


3757/3757 - 1s - loss: 0.9644 - categorical_accuracy: 0.6168 - 1s/epoch - 297us/step


[I 2023-11-03 08:26:18,045] Trial 21 finished with value: 0.6168039441108704 and parameters: {'nlayers': 3, 'nhidden_l0': 180, 'nhidden_l1': 43, 'nhidden_l2': 64, 'learning_rate': 0.00016466746342595515, 'bsize': 128}. Best is trial 16 with value: 0.635113537311554.


3757/3757 - 1s - loss: 0.9292 - categorical_accuracy: 0.6313 - 991ms/epoch - 264us/step


[I 2023-11-03 08:29:06,049] Trial 22 finished with value: 0.6312952041625977 and parameters: {'nlayers': 3, 'nhidden_l0': 170, 'nhidden_l1': 21, 'nhidden_l2': 254, 'learning_rate': 0.00028820287926691784, 'bsize': 128}. Best is trial 16 with value: 0.635113537311554.


3757/3757 - 1s - loss: 0.9218 - categorical_accuracy: 0.6376 - 1s/epoch - 268us/step


[I 2023-11-03 08:32:24,792] Trial 23 finished with value: 0.6375842094421387 and parameters: {'nlayers': 3, 'nhidden_l0': 191, 'nhidden_l1': 30, 'nhidden_l2': 235, 'learning_rate': 0.00027187649873793024, 'bsize': 128}. Best is trial 23 with value: 0.6375842094421387.


3757/3757 - 1s - loss: 0.9829 - categorical_accuracy: 0.6097 - 879ms/epoch - 234us/step


[I 2023-11-03 08:34:13,949] Trial 24 finished with value: 0.6096913814544678 and parameters: {'nlayers': 2, 'nhidden_l0': 116, 'nhidden_l1': 32, 'learning_rate': 0.0002681871285083149, 'bsize': 128}. Best is trial 23 with value: 0.6375842094421387.


3757/3757 - 1s - loss: 0.9652 - categorical_accuracy: 0.6170 - 952ms/epoch - 253us/step


[I 2023-11-03 08:36:24,442] Trial 25 finished with value: 0.6169869303703308 and parameters: {'nlayers': 3, 'nhidden_l0': 201, 'nhidden_l1': 14, 'nhidden_l2': 8, 'learning_rate': 0.00024335131383543718, 'bsize': 128}. Best is trial 23 with value: 0.6375842094421387.


3757/3757 - 1s - loss: 0.9146 - categorical_accuracy: 0.6397 - 999ms/epoch - 266us/step


[I 2023-11-03 08:39:29,203] Trial 26 finished with value: 0.6396722197532654 and parameters: {'nlayers': 3, 'nhidden_l0': 212, 'nhidden_l1': 40, 'nhidden_l2': 167, 'learning_rate': 0.00026892676045040984, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9763 - categorical_accuracy: 0.6136 - 928ms/epoch - 247us/step


[I 2023-11-03 08:41:46,163] Trial 27 finished with value: 0.6136012077331543 and parameters: {'nlayers': 2, 'nhidden_l0': 135, 'nhidden_l1': 39, 'learning_rate': 0.0002718268460212474, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9290 - categorical_accuracy: 0.6339 - 1s/epoch - 282us/step


[I 2023-11-03 08:45:01,390] Trial 28 finished with value: 0.6339239478111267 and parameters: {'nlayers': 3, 'nhidden_l0': 180, 'nhidden_l1': 59, 'nhidden_l2': 148, 'learning_rate': 0.00026991147829656057, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9446 - categorical_accuracy: 0.6226 - 929ms/epoch - 247us/step


[I 2023-11-03 08:51:24,855] Trial 29 finished with value: 0.6225605010986328 and parameters: {'nlayers': 3, 'nhidden_l0': 64, 'nhidden_l1': 57, 'nhidden_l2': 95, 'learning_rate': 0.00029627412614332063, 'bsize': 32}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9778 - categorical_accuracy: 0.6124 - 866ms/epoch - 231us/step


[I 2023-11-03 08:54:28,943] Trial 30 finished with value: 0.6124032735824585 and parameters: {'nlayers': 2, 'nhidden_l0': 105, 'nhidden_l1': 32, 'learning_rate': 0.00024523663472319265, 'bsize': 64}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9350 - categorical_accuracy: 0.6303 - 1s/epoch - 269us/step


[I 2023-11-03 08:57:27,478] Trial 31 finished with value: 0.6302970051765442 and parameters: {'nlayers': 3, 'nhidden_l0': 201, 'nhidden_l1': 21, 'nhidden_l2': 188, 'learning_rate': 0.00026668150993942717, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9259 - categorical_accuracy: 0.6341 - 1s/epoch - 272us/step


[I 2023-11-03 09:00:51,845] Trial 32 finished with value: 0.6341069936752319 and parameters: {'nlayers': 3, 'nhidden_l0': 214, 'nhidden_l1': 18, 'nhidden_l2': 249, 'learning_rate': 0.00023850353676705037, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9514 - categorical_accuracy: 0.6209 - 946ms/epoch - 252us/step


[I 2023-11-03 09:03:31,165] Trial 33 finished with value: 0.6208717823028564 and parameters: {'nlayers': 3, 'nhidden_l0': 143, 'nhidden_l1': 12, 'nhidden_l2': 178, 'learning_rate': 0.00022313254764708663, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9201 - categorical_accuracy: 0.6367 - 1s/epoch - 269us/step


[I 2023-11-03 09:06:42,239] Trial 34 finished with value: 0.6367107629776001 and parameters: {'nlayers': 3, 'nhidden_l0': 224, 'nhidden_l1': 17, 'nhidden_l2': 179, 'learning_rate': 0.00027643753548094294, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9295 - categorical_accuracy: 0.6324 - 989ms/epoch - 263us/step


[I 2023-11-03 09:09:53,503] Trial 35 finished with value: 0.6323516964912415 and parameters: {'nlayers': 3, 'nhidden_l0': 161, 'nhidden_l1': 29, 'nhidden_l2': 166, 'learning_rate': 0.00027993372011527666, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9755 - categorical_accuracy: 0.6164 - 989ms/epoch - 263us/step


[I 2023-11-03 09:12:33,771] Trial 36 finished with value: 0.6164129376411438 and parameters: {'nlayers': 2, 'nhidden_l0': 252, 'nhidden_l1': 23, 'learning_rate': 0.00025916565596328706, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9349 - categorical_accuracy: 0.6332 - 1s/epoch - 275us/step


[I 2023-11-03 09:16:23,105] Trial 37 finished with value: 0.6332252025604248 and parameters: {'nlayers': 3, 'nhidden_l0': 196, 'nhidden_l1': 73, 'nhidden_l2': 96, 'learning_rate': 0.0002993423610613718, 'bsize': 128}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9229 - categorical_accuracy: 0.6365 - 971ms/epoch - 258us/step


[I 2023-11-03 09:23:16,320] Trial 38 finished with value: 0.6365194320678711 and parameters: {'nlayers': 3, 'nhidden_l0': 129, 'nhidden_l1': 36, 'nhidden_l2': 194, 'learning_rate': 0.0002518607304828508, 'bsize': 32}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9579 - categorical_accuracy: 0.6238 - 913ms/epoch - 243us/step


[I 2023-11-03 09:28:39,953] Trial 39 finished with value: 0.6238083243370056 and parameters: {'nlayers': 2, 'nhidden_l0': 120, 'nhidden_l1': 35, 'learning_rate': 0.00025367809359018597, 'bsize': 32}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9391 - categorical_accuracy: 0.6259 - 938ms/epoch - 250us/step


[I 2023-11-03 09:35:20,848] Trial 40 finished with value: 0.6259129643440247 and parameters: {'nlayers': 3, 'nhidden_l0': 97, 'nhidden_l1': 37, 'nhidden_l2': 177, 'learning_rate': 0.000228575373251494, 'bsize': 32}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9661 - categorical_accuracy: 0.6194 - 909ms/epoch - 242us/step


[I 2023-11-03 09:40:25,537] Trial 41 finished with value: 0.6193661093711853 and parameters: {'nlayers': 2, 'nhidden_l0': 128, 'nhidden_l1': 28, 'learning_rate': 0.00024254458998685524, 'bsize': 32}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9228 - categorical_accuracy: 0.6364 - 957ms/epoch - 255us/step


[I 2023-11-03 09:47:14,062] Trial 42 finished with value: 0.6363613605499268 and parameters: {'nlayers': 3, 'nhidden_l0': 155, 'nhidden_l1': 19, 'nhidden_l2': 189, 'learning_rate': 0.00028170916328521244, 'bsize': 32}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9188 - categorical_accuracy: 0.6387 - 994ms/epoch - 265us/step


[I 2023-11-03 09:54:14,648] Trial 43 finished with value: 0.6386823058128357 and parameters: {'nlayers': 3, 'nhidden_l0': 213, 'nhidden_l1': 19, 'nhidden_l2': 127, 'learning_rate': 0.00027641823129673595, 'bsize': 32}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9200 - categorical_accuracy: 0.6389 - 1s/epoch - 291us/step


[I 2023-11-03 10:01:13,249] Trial 44 finished with value: 0.6388986110687256 and parameters: {'nlayers': 3, 'nhidden_l0': 209, 'nhidden_l1': 17, 'nhidden_l2': 129, 'learning_rate': 0.0002605845976699137, 'bsize': 32}. Best is trial 26 with value: 0.6396722197532654.


3757/3757 - 1s - loss: 0.9107 - categorical_accuracy: 0.6408 - 998ms/epoch - 266us/step


[I 2023-11-03 10:08:13,435] Trial 45 finished with value: 0.6408368945121765 and parameters: {'nlayers': 3, 'nhidden_l0': 214, 'nhidden_l1': 16, 'nhidden_l2': 125, 'learning_rate': 0.0002723043751349653, 'bsize': 32}. Best is trial 45 with value: 0.6408368945121765.


3757/3757 - 1s - loss: 0.9028 - categorical_accuracy: 0.6446 - 1s/epoch - 266us/step


[I 2023-11-03 10:15:06,762] Trial 46 finished with value: 0.6445720195770264 and parameters: {'nlayers': 3, 'nhidden_l0': 201, 'nhidden_l1': 14, 'nhidden_l2': 124, 'learning_rate': 0.0002644115415668436, 'bsize': 32}. Best is trial 46 with value: 0.6445720195770264.


3757/3757 - 1s - loss: 1.0896 - categorical_accuracy: 0.5739 - 925ms/epoch - 246us/step


[I 2023-11-03 10:20:52,631] Trial 47 finished with value: 0.5738707184791565 and parameters: {'nlayers': 1, 'nhidden_l0': 222, 'learning_rate': 0.00023145994100466462, 'bsize': 32}. Best is trial 46 with value: 0.6445720195770264.


3757/3757 - 1s - loss: 0.9061 - categorical_accuracy: 0.6466 - 1s/epoch - 278us/step


[I 2023-11-03 10:28:02,519] Trial 48 finished with value: 0.6465518474578857 and parameters: {'nlayers': 3, 'nhidden_l0': 246, 'nhidden_l1': 14, 'nhidden_l2': 125, 'learning_rate': 0.0002518045355374134, 'bsize': 32}. Best is trial 48 with value: 0.6465518474578857.


3757/3757 - 1s - loss: 0.9084 - categorical_accuracy: 0.6441 - 995ms/epoch - 265us/step


[I 2023-11-03 10:35:17,577] Trial 49 finished with value: 0.6441227793693542 and parameters: {'nlayers': 3, 'nhidden_l0': 228, 'nhidden_l1': 14, 'nhidden_l2': 104, 'learning_rate': 0.00026048575367957103, 'bsize': 32}. Best is trial 48 with value: 0.6465518474578857.


3757/3757 - 1s - loss: 0.9169 - categorical_accuracy: 0.6389 - 1s/epoch - 274us/step


[I 2023-11-03 10:42:06,177] Trial 50 finished with value: 0.6389152407646179 and parameters: {'nlayers': 3, 'nhidden_l0': 243, 'nhidden_l1': 13, 'nhidden_l2': 96, 'learning_rate': 0.00021594626953928036, 'bsize': 32}. Best is trial 48 with value: 0.6465518474578857.


3757/3757 - 1s - loss: 0.9193 - categorical_accuracy: 0.6378 - 947ms/epoch - 252us/step


[I 2023-11-03 10:48:49,628] Trial 51 finished with value: 0.6378005146980286 and parameters: {'nlayers': 3, 'nhidden_l0': 178, 'nhidden_l1': 10, 'nhidden_l2': 107, 'learning_rate': 0.00023883917025148635, 'bsize': 32}. Best is trial 48 with value: 0.6465518474578857.


Study statistics: 
  Number of finished trials:  52
  Number of trials:  52
Best trial:
  Value:  0.6465518474578857
  Params: 
    nlayers: 3
    nhidden_l0: 246
    nhidden_l1: 14
    nhidden_l2: 125
    learning_rate: 0.0002518045355374134
    bsize: 32


## MLP 32:
## BEST  Trial 48 finished with value: 0.6465518474578857 and parameters: {'nlayers': 3, 'nhidden_l0': 246, 'nhidden_l1': 14, 'nhidden_l2': 125, 'learning_rate': 0.0002518045355374134, 'bsize': 32}

## MLP 16 : 
##  BEST Trial 48 finished with value:0.655561089515686 and parameters: {'bsize': 64, 'nlayers': 3, 'nhidden_l0': 171, 'nhidden_l1': 30, 'nhidden_l2': 153, 'learning_rate': 0.00025664296269595563}.

## MLP 8:
## BEST Trial 47 finished with value: 0.6322103142738342 and parameters: {'bsize': 64, 'nlayers': 3, 'nhidden_l0': 152, 'nhidden_l1': 69, 'nhidden_l2': 63, 'learning_rate': 0.00023812077479493001}